src/dex/swaap-v1/swaap-v1.ts

```diff
@@ -0,0 +1,633 @@
+import { SimpleExchange } from '../simple-exchange';
+import { SwaapV1Config, Adapters } from './config';
+import { SwaapV1Pool } from './swaap-v1-pool';
+import PoolABI from '../../abi/swaap-v1/pool.json';
+import { PoolQuotations } from './libraries/PoolQuotations';
+
+const subgraphTimeout = 1000 * 10;
+const MAX_POOL_CNT = 1000;
+const POOL_CACHE_TTL = 60 * 60; // 1hr
+
+const fetchAllPools = `query ($count: Int) {
+  pools: pools(first: $count, orderBy: liquidity, orderDirection: desc, where: {finalized: true}) {
+    id
+    tokens {
+      address
+      decimals
+      oracleInitialState {
+        proxy
+        price: fixedPointPrice
+        decimals
+      }
+    }
+    liquidityUSD: liquidity
+    swapFee
+    dynamicCoverageFeesZ
+    dynamicCoverageFeesHorizon
+    priceStatisticsLookbackInSec
+    priceStatisticsLookbackInRound
+    priceStatisticsLookbackStepInRound
+    maxPriceUnpegRatio
+  }
+}`;
+
+export class SwaapV1 extends SimpleExchange implements IDex<SwaapV1Data> {
+  static readonly poolInterface = new Interface(PoolABI);
+
+  protected config: DexParams;
+  protected subgraphURL: string;
+  protected exchangeProxy: string;
+
+  protected eventPools: { [poolAddress: string]: SwaapV1Pool } = {};
+
+  public allPools: Map<string, SubgraphPoolBase> = new Map();
+
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly isFeeOnTransferSupported: boolean = false;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(SwaapV1Config);
+
+  logger: Logger;
+
+  constructor(
+    protected network: Network,
+    protected dexKey: string,
+    protected dexHelper: IDexHelper,
+    protected adapters = Adapters[network], // TODO: add any additional optional params to support other fork DEXes
+  ) {
+    super(dexHelper.config.data.augustusAddress, dexHelper.web3Provider);
+    this.config = SwaapV1Config[dexKey][network];
+    this.logger = dexHelper.getLogger(`${dexKey}-${network}`);
+    this.subgraphURL = this.config.subgraphURL;
+    this.exchangeProxy = this.config.exchangeProxy;
+  }
+
+  async init(blockNumber: number) {
+    this.allPools = await this.getAllPools(blockNumber);
+  }
+
+  // Initialize pricing is called once in the start of
+  // pricing service. It is intended to setup the integration
+  // for pricing requests. It is optional for a DEX to
+  // implement this function
+  async initializePricing(blockNumber: number) {
+    await this.init(blockNumber);
+    if (!this.allPools)
+      throw new Error(
+        'initializePricing: SwaapV1 cfgInfo still null after init',
+      );
+    for (const poolConfig of this.allPools.values()) {
+      await this.addPool(poolConfig, blockNumber);
+    }
+  }
+
+  async addPool(poolConfig: SubgraphPoolBase, blockNumber: number) {
+    const pool = new SwaapV1Pool(
+      this.dexKey,
+      this.network,
+      poolConfig,
+      this.dexHelper,
+    );
+    const state = await pool.generateState(blockNumber);
```

Comment:
I would add an optional parameter state to this function, and I will call generateState if it's not provided. This will permit you to batch the initial setState but fallback to generateState when a new pool is added

Reply:
not sure to understand what would be the use case of "This will permit you to batch the initial setState but fallback to generateState when a new pool is added"?

Reply:
Related to https://github.com/paraswap/paraswap-dex-lib/pull/154#discussion_r1004688618
You can ignore it now :)